Hide redundant parameters, add shipParameters to levelData, Improve shipParameter legacy suport, & gost fixes

CardamomCake opened the pull request:
This pr also implements boostSpoolUpTime in the GUI, and makes sure it correctly during the countdown. However, audio is unaffected for now, thus this variable is also hidden from the menu like mass. 

It also makes sure that unspecified ship parameters are set to default when loading them from a json. 

I hope that you can simply slot in the "Dev Pannel.prefab" file without issue, but apparently, this can create problems. For now it is included in this pr.

CardamomCake commented:
I found I made a mistake which caused the loading shipConfig from clipboard always set it to default parameters.

jukibom commented:
@CardamomCake what was the rationale for removing Mass again? I find it easier to e.g. reduce the mass than to increase the thrust when messing about because the the thrust applies to other multipliers

jukibom commented on Assets/Scripts/Core/ShipModel/ShipPhysics.cs:
```diff
@@ -575,8 +575,8 @@ public class ShipPhysics : MonoBehaviour {
                     if (latInputTotal > 0) {
-                        var latHComponent = Mathf.Abs(latH) / latInputTotal;
-                        var latVComponent = Mathf.Abs(latV) / latInputTotal;
+                        var latHComponent = Mathf.Abs(latH) / latInputTotal * FlightParameters.boostDivertEfficiency;
```
so here we're doing the division first then the multiplication and only if the input is greater than 1 so there's a couple things I don't quite get here

* if boostDivertEfficiency is greater than 1 we gain extra thrust from nowhere? (previously this logic steals boost from forward thrust by diverting it into lats). Is this value intended to be less than 1 or left open-ended?
* if you _only_ hold, say, right but not up or down then this logic won't apply anyway right?

CardamomCake replied:
- |if boostDivertEfficiency is greater than 1 we gain extra thrust from nowhere? (previously this logic steals boost from forward thrust by diverting it into lats). Is this value intended to be less than 1

Its intended to be less than one, but if other people want it higher in order to make a crazy bouncy build, I don't see the harm. 

- if you only hold, say, right but not up or down then this logic won't apply anyway right?

If any of the lats are nonzero, the logic will apply (line 576, 577), but only to the degree that that specific lateral utilized. Boosting with only using the right direction will result in the latVCompenent being multiplied with zero, and the latHComponent being multiplied with something positive.  It is just your old logic but with an extra multiplier to control the magnitude of the effect. The throttle part of the logic is more difficult, but I didn't touch that. I think your logic is good, but it is quite a mind bender with the whole throttle remap thing. If it is wrong, I think we're screwed because I assume this is the logic the current flight system and all the times are based on. I am positive that my change only does what is advertised, tho.

jukibom commented on Assets/Scripts/Core/ShipModel/ShipParameters.cs:
```diff
@@ -60,24 +65,281 @@ public class ShipParameters {
+
+                var parameters = JsonConvert.DeserializeObject<NullableParameters>(json);
+
+                if (parameters.angularDrag != Defaults.angularDrag)
```
there looked to be something much cleaner and simpler here before, what happened? 

this looks like a lot of duplicated logic! I wonder if that can be combined into a function... (I have no idea how to reference object variable names as parameters but I'm almost certain there'll be a way).

Alternatively, I think you can mark json fields in the object with [Default] attributes or something to do similar logic to this (IIRC this is why exported level json omits zero values on coordinates, the default attribute just ignores them)

CardamomCake replied:
I apologize for this mess, the original code turned out to be completely broken, where the only reason it didnt crash was because it didn't run at all.  I probably should not have submitted this mess, but I mostly work on this when it is late, and I wanted it fixed and sleep. 

I am working on something much cleaner now, but that also is giving me some problems for now. Il try to get that submitted.

jukibom commented on Assets/Scripts/Menus/Options/DevPanel.cs:
```diff
@@ -96,13 +114,30 @@ public class DevPanel : MonoBehaviour {
+        var json = GetFlightParams().ToJsonString();
+
+        var jObject = JObject.Parse(json);
+        if ((float)jObject.GetValue("mass") == ShipParameters.Defaults.mass)
```
if the defaults attribute thing mentioned above was implemented, I _think_ these could be removed (nothing would export unless different to default)

jukibom commented on Assets/Scripts/Menus/Options/DevPanel.cs:
```diff
@@ -96,13 +114,30 @@ public class DevPanel : MonoBehaviour {
+            jObject.Remove("mass");
+        if ((float)jObject.GetValue("inertiaTensorMultiplier") == ShipParameters.Defaults.inertiaTensorMultiplier)
+            jObject.Remove("inertiaTensorMultiplier");
+        if ((float)jObject.GetValue("boostSpoolUpTime") == ShipParameters.Defaults.boostSpoolUpTime) // again remove if boost spool up time has been properly implemented. 
```
okay so to be clear this value is added in the json but omitted from export right now and not used because of audio?

CardamomCake replied:
Yes, it is also hidden from the menu.

CardamomCake commented:
Il close it till got that flight parameter mess fixed

CardamomCake commented:
Well, it is a bit of a large pull request now. 

- Hide mass and inertiaTensorMultiplier unless they are set differently from default 
- Loading shipParameters now sets all the missing parameters to default instead to zero.
- Changed shipParameters.Defaults to shipParameter.createDefaults to work around an issue of the json parser. And reworked how defaults are handled in general. 
- Added shipParameters to the levelData so it can be loaded via json. 
- Fixed an issue with ghosts where they did not load their flightParameters
- Fixed an issue with ghosts where they were not loaded when loading a custom race
